src/dht_experimental.rs

```diff
@@ -0,0 +1,67 @@
+use anyhow::{Context, Result};
+use ed25519_dalek::VerifyingKey;
+
+use futures_lite::StreamExt;
+use mainline_exp::{Dht as MainlineDht, Id, SigningKey, async_dht::AsyncDht};
+use sha2::Digest;
+
+#[derive(Debug, Clone)]
+pub struct Dht {
+    dht: Option<AsyncDht>,
+    signing_key: SigningKey,
+}
+
+impl Dht {
+    pub fn new(signing_key: &SigningKey) -> Self {
+        Self {
+            dht: None,
+            signing_key: signing_key.clone(),
+        }
+    }
+
+    pub fn reset(&mut self) -> Result<()> {
+        self.dht = Some(MainlineDht::builder().build()?.as_async());
+        Ok(())
+    }
+
+    pub async fn get_peers(&mut self, topic_bytes: &Vec<u8>) -> Result<Vec<VerifyingKey>> {
+        if self.dht.is_none() {
+            self.reset()?;
+        }
+        let mut hasher = sha2::Sha512::new();
+        hasher.update(topic_bytes);
+        let hash: [u8; 20] = hasher.finalize()[..20]
+            .try_into()
+            .context("Failed to convert hash")?;
+
+        let dht = self.dht.as_mut().context("DHT not initialized")?;
+        let id = Id::from_bytes(hash)?;
+
+        let topic_stream = dht.get_signed_peers(id).await.collect::<Vec<_>>();
+        Ok(topic_stream
+            .await
+            .iter()
+            .flatten()
+            .filter_map(|item| VerifyingKey::from_bytes(item.key()).ok())
+            .collect::<Vec<_>>())
+    }
+
+    pub async fn announce_self(&mut self, topic_bytes: &Vec<u8>) -> Result<()> {
+        if self.dht.is_none() {
+            self.reset()?;
+        }
+        let mut hasher = sha2::Sha512::new();
+        hasher.update(topic_bytes);
```

Comment:
It might be useful here to add a namespace like `b'Iroh'` just to make sure that anyone using the same topic and sha512 but for another overlay network than Iroh, won't get the same `info_hash`.. I described this in the BEP, but maybe I should have added it to the function signature? I didn't want to force it on people.